Add enum type to rpc registry for consolidating RPC initialization code path

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28628 Add enum type to rpc registry for consolidating RPC initialization code path**

Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

xush6528 committed:
Add enum type to rpc registry for consolidating RPC initialization code path

Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

mrshenli commented on test/dist_autograd_test.py:
```diff
@@ -258,6 +258,15 @@ def _verify_graph_for_nested_rpc_call(self, ctx):
     def _test_graph(self, fn):
         dst_rank = (self.rank + 1) % self.world_size
+
+        if not dist.is_initialized():
```
Let's add a comments here to mention that this is needed because this test case calls `dist.barrier()` in the end.

xush6528 replied:
Added.

pietern replied:
Too much duplication. Consider adding a `with run_barrier_on_exit()`, or a test decorator, for example.

mrshenli commented on test/rpc_test.py:
```diff
@@ -261,53 +264,29 @@ def test_duplicate_name(self):
         rpc.join_rpc()
 
-    @dist_init(setup_model_parallel=False)
-    def test_init_invalid_backend(self):
```
why removing this test?

xush6528 replied:
Because we are changing to Python Enum class. 

`backend_name` string got tested by Enum class internals, which is a Python standard library thing. No need to test Python std library.

mrshenli commented on test/rpc_test.py:
```diff
@@ -243,16 +252,10 @@ def test_register_rpc_backend_and_start_rpc_backend(
     @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
     @dist_init(setup_model_parallel=False)
     def test_duplicate_name(self):
-        dist.init_process_group(
```
Why this test does not require a pg? ProcessGroupAgent would ignore `requires_process_group_agent` and still need the the pg instance to collect names and check if they match, no?

xush6528 replied:
Because in this PR, I move the pg initialized inside of `def _init_rpc` instead, `def process_group_init_backend_handler` to be precise.

No, I changed `@requires_process_group_agent` to test agianst test config instead of the current agen instance that has been initialzied.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -21,14 +26,56 @@ def register_backend(backend_name, init_backend_handler):
+    # Create a new enum type, `BackendType`, with extended members.
+    global BackendType
+    existing_enum_dict = {member.name: member.value for member in BackendType}
+    extended_enum_dict = dict(
```
why do we need to have two dicts here instead of directly inserting the new backend?

xush6528 replied:
Becuase Enum type is a immutable datastrucuture. There is no API from the std library to insert a new enum variant. I have todeclare a new custom Enum class dynamically.

mrshenli replied:
I was talking about have two dicts here. I see.

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -1,15 +1,20 @@
-def is_backend_registered(backend_name):
-    return backend_name in _get_backend_registry()
+
+def _is_backend_name_registered(backend_name):
```
Is this function just used once in `register_backend`? If that's the case, would it be better to make it inline?

xush6528 replied:
Yes, it's not an external API anymore. `backend_name` validity is tested when accessing a particular invariont of the enum type.

For this short one, it's ok to make it inline.

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 commented:
Address @mrshenli 's comments

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -21,14 +22,56 @@ def register_backend(backend_name, init_backend_handler):
+    # Initialize ProcessGroup.
+    if not dc10d.is_initialized():
+        world_size = len(worker_name_to_id)
+        dc10d.init_process_group(
```
This is a side effect of this method. Shall we add docs to `init_model_parallel` to describe this?

mrshenli replied:
I feel a better solution might be letting ProcessGroupAgent to have its own pg instance instead of modifying the global one. But I am OK to land this as is to unblock.

xush6528 replied:
Yes, this will caues confusion to users. We could address this later. 

The work comes from the fact that all helper functions in `distributed_c10d.py` modify the global varaiables like `_default_pg` and `_pg_map`. We need to refactor distributed_c10d.py` a bit to have basic building block that only returns a `pg` withought manipulate the global variables. Then, put a `_model_parallel_pg` in `model_parallel/__init__.py`

mrshenli commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -21,14 +22,56 @@ def register_backend(backend_name, init_backend_handler):
+            backend="gloo", store=store, rank=self_rank, world_size=world_size
+        )
+
+    group = dc10d._get_default_group()
```
Is this the only function that's not available in `torch.distributed` and used in this file? If yes, shall we import this function explicitly from `distributed_c10d`. The other two functions `is_initialized` and `init_process_group` can be called from `torch.distributed`.

xush6528 replied:
Updated. Only this function is called through `distributed_c10d` module.

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -43,14 +43,22 @@ def init_model_parallel(
         """
+        # Rendezvous.
+        world_size = len(worker_name_to_id)
+        rendezvous_iterator = torch.distributed.rendezvous(
```
I recall that we ran into an issue of rendezvous twice when using `ProcessGroupAgent` when doing it here. Wouldn't this hit that problem again?

xush6528 replied:
I do handle this by inserting a few process_group barrier to circumvent it. 
Putting barrier between failed rendezvous versions could avoid mis-match.

pietern replied:
What do you mean here? Mismatch between torch.distributed and torch.distributed.rpc?

xush6528 replied:
Mismatch between multiple calls to `rendezvous`, an issue described in #28552.

mrshenli reviewed (approved):
Stamp to unblock. Please address test failure and comments before landing.

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 commented:
`test_invalid_names` failed due to the reason as described in #28552.
Adding a temporary hack to circumvent it.

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

xush6528 committed:
Update on "Add enum type to rpc registry for consolidating RPC initialization code path"


Consolidate code paths of ProcessGroupAgent construction and other RPC Backend construction.

Differential Revision: [D5516188](https://our.internmc.facebook.com/intern/diff/D5516188/)

[ghstack-poisoned]

pietern commented on test/dist_autograd_test.py:
```diff
@@ -702,7 +760,7 @@ def test_backward_node_failure(self):
             if self.rank % 2 == 0:
                 # Wait a bit for all other nodes to die.
-                time.sleep(5)
+                time.sleep(5)  # This is flaky.
```
Please add an issue to fix this and link it in the skip description.

xush6528 replied:
Added in #28977.

pietern commented on test/dist_utils.py:
```diff
@@ -51,25 +50,29 @@ def set_termination_signal():
 
 
-def dist_init(test_method=None, setup_model_parallel=True, clean_shutdown=True):
+def dist_init(old_test_method=None, setup_model_parallel=True, clean_shutdown=True):
```
Why the rename?

xush6528 replied:
Not fully intended, I tink I picked #27940 instead of #27940 when resolving conflicts.

pietern replied:
I see. I don't have a preference, but don't like anything with `old` or `new` in it, because it implies something is either being introduced or being deprecated. And for these decorators it's not needed if you use the functools wrappers.

pietern commented on test/rpc_test.py:
```diff
@@ -261,87 +265,82 @@ def test_duplicate_name(self):
 
+        base_file_name = self.file_name
+        # Use a different file path for FileStore to avoid rendezvous mismatch.
+        self.file_name = base_file_name + "1"
```
Will this file be removed from disk?

xush6528 replied:
None of the temp file is removed.

https://github.com/pytorch/pytorch/blob/6eef4690743a31ccc546f0c4f2d057d602e1e79e/test/common_distributed.py#L155

pietern commented on torch/distributed/rpc/backend_registry.py:
```diff
@@ -21,14 +22,65 @@ def register_backend(backend_name, init_backend_handler):
+        **existing_enum_dict
+    )
+    BackendType = enum.Enum(value="BackendType", names=extended_enum_dict)
+    return BackendType[backend_name]
```
Not a fan of this approach. It means that `BackendType.FOO_BAR` might fail depending on when you call it.

What's wrong with using a string? Common strings can still be exposed as constant.

xush6528 replied:
I intend to fail in that case.
The failure would be a clear and standalized Python built-in enum module "no attribute" error, clearly indicating users that they should register their custom backend type into the type enum first.

pietern reviewed (commented):
There is a lot to unpack in this PR. I feel it should have been a stack of 3 or 4 PRs instead of one big one.

The title lets me know what was done (the enum type), but not why.

What is the reason for using an enum here?

xush6528 commented:
@pietern 

```py
WorkerCtx.backend() is rpc.backend_registry.BackendType.PROCESS_GROUP
```
The benifit of using enum for this use is it's object equality test instead of string comparison.

rohan-varma commented on test/rpc_test.py:
```diff
@@ -363,7 +362,6 @@ def test_invalid_names(self):
         # would add extra overhead to the call, and normal use cases won't
         # create a progress group and exit without doing anything. Hence, it is
         # not worthy to introduce the overhead just for this test case.
-        dist.barrier()
```
Any reason this was removed? @mrshenli added it in https://github.com/pytorch/pytorch/pull/28376, to fix this test's flakiness.

xush6528 replied:
@rohan-varma Since the one successful `dist.init_process_group(..)` on line 302 was removed too, there is no need to wait for all workers to be synced.

mrshenli replied:
as RPC is creating a PG internally, it seems we need to call barrier on that internal PG? @xush6528 @rohan-varma

xush6528 replied:
@mrshenli We discussed offline,
since "Connection closed by peer" error should not be thrown from `ProcessGroupAgent` constructor,  the problem might be the `destroy_process_group()` call in the python `except Exception:` section in , `process_group_init_backend_handler(..)`, which is called by `_init_rpc_backend(...)`.
